NVStore: add the allocate_key API (instead of set_alloc_key)

davidsaada opened the pull request:
### Description
Add the allocate_key API. This replaces the previously added set_alloc_key API (which allocates a key and sets the value at the same time). 
Reason for the change: Key allocation will typically be used by other storage features (like StorageLite), keeping the allocated keys in another location. Previous API created problems in the case key allocation and value setting couldn't be done at the same time (for instance, if the set value was derived from the allocated key, such as hash or CMAC). 

### Pull request type

[ ] Fix  
[ ] Refactor  
[ ] New target  
[ ] Feature  
[x] Breaking change

Commit:
NVStore: add the allocate_key API (instead of set_alloc_key)

Add the allocate_key API. This replaces the previously added set_alloc_key API
(which allocates a key and sets the value at the same time).
Reason for the change: Key allocation will typically be used by other storage
features (like StorageLite), keeping the allocated keys in another location.
Previous API created problems in the case key allocation and value setting
couldn't be done at the same time (for instance, if the set value was
derived from the allocated key, such as hash or CMAC).

theotherjimmy commented on features/nvstore/source/nvstore.h:
```diff
@@ -159,22 +159,14 @@ class NVStore : private mbed::NonCopyable<NVStore> {
      *
      */
-    int set_alloc_key(uint16_t &key, uint16_t buf_size, const void *buf);
+    int allocate_key(uint16_t &key);
```
This is a breaking change.

davidsaada replied:
It is indeed, but the replaced function (`set_alloc_key`) is a new one that was implemented after the 5.8 release. It's still not even documented. Figured that the API we had defined was wrong in that case, hence the change. Can add the "old" API back, but that that would be just dead code.

theotherjimmy replied:
This PR contains a breaking change, so marked for 5.9.0.

davidsaada replied:
That's OK. It is mainly used for StorageLite feature, which should be out in 5.9.0.

theotherjimmy commented on features/nvstore/source/nvstore.h:
```diff
@@ -422,14 +414,14 @@ class NVStore : private mbed::NonCopyable<NVStore> {
      * @returns 0 for success, nonzero for failure.
      */
-    int do_set(uint16_t &key, uint16_t buf_size, const void *buf, uint16_t flags);
+    int do_set(uint16_t key, uint16_t buf_size, const void *buf, uint16_t flags);
```
This is also a breaking change.

davidsaada replied:
That's not a breaking change, as the function is private (maybe the doxy comments are confusing to think it is not).

theotherjimmy replied:
Ah, I missed the private at the top. Thanks.

0xc0170 commented on features/nvstore/source/nvstore.cpp:
```diff
@@ -597,16 +599,16 @@ int NVStore::do_get(uint16_t key, uint16_t buf_size, void *buf, uint16_t &actual
     _mutex->lock();
-    record_offset = _offset_by_key[key];
+	
+	record_offset = _offset_by_key[key];
```
misaligned- tabs here?

davidsaada replied:
Strange, my editor is setup to insert spaces only. Anyway, fixed it.

0xc0170 reviewed (changes requested):
LGTM, just one small change

0xc0170 reviewed (changes requested):
new APi additions should contain better commit messages

SenRamakri reviewed (dismissed):
Looks good.

davidsaada commented:
Changed the `allocate_key` API, adding an optional parameter of owner, and added another API - ` free_all_keys_by_owner`
This was added due to some feedbacks from features using `allocate_key`: Once the keys are allocated, it is hard to keep track of them. This is specially true when we want to perform an initialization. We would like to remove all dynamically allocated keys belonging to the feature as part of the initialization process. Therefore, an optional `owner` parameter was added to the `allocate_key` API (if omitted, key will have an unspecified owner). owner here is a fixed, predefined feature, allowing us now to remove all keys using the ` free_all_keys_by_owner` API.
This change depends on PR [#6864](https://github.com/ARMmbed/mbed-os/pull/6864) - only because that PR fixes some failures in NVStore tests.

davidsaada commented:
Rebased following merge of PR [#6864](https://github.com/ARMmbed/mbed-os/pull/6864). No longer depending on that PR.

cmonr commented:
/morph build

cmonr commented:
/morph test

cmonr commented:
/morph export-build

cmonr commented:
/morph test

cmonr commented:
/morph test

adbridge commented:
@SenRamakri @0xc0170 Are you happy with the updates ?

cmonr commented:
Looks like @0xc0170 was already good with the changes.
